apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java

```diff
@@ -32,209 +65,403 @@
+   * ALWAYS >= the number of jobs that are actually running, from the view of the {@link
+   * SchedulerJob}'s thread. This aligns with what the {@link SchedulerJob} needs the value for, as
+   * we're using it to avoid over-subscribing on work. Under-subscribing is fine, in the context.
+   *
+   * @return the counter for how many tasks are currently running on the {@link PipelineJob}s {@link
+   *     Executor}
+   */
+  private final AtomicInteger jobsRunningCounter;
+
+  /**
+   * This {@link Set} stores handles to all active job {@link Future}s. We need to keep track of
+   * these so that we can ensure that jobs can be cancelled when {@link #stop()} is called. It needs
+   * to be thread-safe because, to prevent runaway memory growth, we remove jobs from the set as
+   * they finish.
+   */
+  private final ConcurrentMap<PipelineJobRecordId, PipelineJobHandle<?>> jobActiveHandles;
+
+  /**
+   * Used to run the job monitoring tasks, e.g. {@link ListenableFuture} callbacks. Those tasks are
+   * run in a separate thread pool to ensure that job execution doesn't starve them out.
+   */
+  private final ExecutorService jobMonitorsExecutor;
+
+  /**
+   * The {@link PipelineJobRecordStore} service that tracks job submissions, executions, and
+   * outcomes.
+   */
+  private final PipelineJobRecordStore jobRecordStore;
+
+  /**
+   * Stores the {@link PipelineJob}s that have been registered via {@link
+   * #registerJob(PipelineJob)}.
+   */
+  private final Map<PipelineJobType<?>, PipelineJob<?>> jobRegistry;
 
   /**
    * Constructs a new {@link PipelineManager} instance. Note that this must be used as a singleton
    * service in the application: only one instance running at a time is supported.
    *
    * @param appMetrics the {@link MetricRegistry} for the overall application
-   * @param options the {@link ExtractionOptions} to use
-   * @param scanRepeatDelay the number of milliseconds to wait after completing one poll/process
-   *     operation and starting another
-   * @param listener the {@link DataSetMonitorListener} that will be notified when events occur
-   */
-  public PipelineManager(
-      MetricRegistry appMetrics,
-      ExtractionOptions options,
-      int scanRepeatDelay,
-      DataSetMonitorListener listener) {
+   * @param jobRecordStore the {@link PipelineJobRecordStore} for the overall application
+   */
+  public PipelineManager(MetricRegistry appMetrics, PipelineJobRecordStore jobRecordStore) {
     this.appMetrics = appMetrics;
-    this.options = options;
-    this.scanRepeatDelay = scanRepeatDelay;
-    this.listener = listener;
+    this.jobRecordStore = jobRecordStore;
+    this.jobExecutor = createJobExecutor();
+    this.jobActiveHandles = new ConcurrentHashMap<>();
+    this.jobsRunningCounter = new AtomicInteger(0);
+    this.jobMonitorsExecutor = Executors.newCachedThreadPool();
+    this.jobRegistry = new HashMap<>();
```

Comment:
jobRegistry needs to be ConcurrentHashMap since we are using stream() in various methods.  Raw HashMap will throw ConcurrentModificationException if anything is added to or removed from the map during iteration/streaming over it.

Reply:
I agree that's the case now and likely to remain the case in the future.  But someone less familiar with the code and unaware of the distinction could come along later and add a call to registerJob() or make some other change without knowing the risk.

Reply:
I understand the concern, but I think it's low risk, just due to how the application works: hard to see why someone would lazy-/late-register jobs, rather than doing so in `PipelineApplication`.

Regardless, though, if concurrency is a concern, then just making the collection concurrent likely wouldn't be enough. You can end up with all sorts of goofy race conditions, even with a "thread safe" collection (as evidenced by the recent commits here). I think, if we want to address it, the right fix is to ensure that the registration API isn't visble to anything other than `PipelineApplication` -- probably by moving jobs into the constructor. Which... is a good idea, actually, and something we should consider doing in the future. This whole thing just went through so many iterations that there are still some goofy bits like that hanging around.